### ISSUE TITLE
Runahead limiting -  behaviour change, flow doesn't stall

### ISSUE THREAD

**Issue by MetRonnie** [cylc/cylc-flow#3792]
**Describe the bug**

Workflow does not stall even if `max active cycle points` or `runahead limit` are set.

**Release version(s) and/or repository branch(es) affected?**

master branch (8.0a3.dev) @ d3bb21da2507913c45086809c63dfaae6fcab166

**Steps to reproduce the bug**

flow.cylc:
```ini
[scheduling]
    cycling mode = integer
    # runahead limit = P2
    max active cycle points = 2
    [[graph]]
        P1 = foo => bar
[runtime]
    [[root]]
        script = true
    [[foo]]
        script = false
```
The suite/flow doesn't stall, it just keeps starting new cycles and spawning new `foo.n`s.

**Expected behavior**

`foo.1` and `foo.2` fail, then the flow stalls. 

**Additional context**

When run on 8.0a2, `bar.1` and `bar.2` spawn but don't run, and then the flow stalls. On master, no `bar` ever spawns.

I'm guessing this is due to SoD, so tagging as such.

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

**Comment by hjoliver** [comment]
@MetRonnie - this really isn't a bug, it's a choice. 

Now that it's possible to do this properly thanks to SoD, I think a flow should only stall if there is literally nothing else it can run.  In this case, if  `foo.1` and `foo.2` fail, and `foo.3` does not depend on them succeeding, then why shouldn't we run it?

So, that being the case, the current runahead limiting ignores failed tasks - it just (deliberately) limits the spread of active tasks over cycle points.

I'm aware that this is a choice I made during SoD implementation, and as such it might require some discussion before we settle which way to go, but it makes sense to me - why should the flow stall in this situation?  

Note we didn't really have a choice pre-SoD, because the spawn-before-needed system resulted in waiting tasks downstream of the failed tasks, and those were "stuck" never going to run.

**Comment by MetRonnie** [comment]
Ok that makes sense.

And I see that if you instead do
```
    [[graph]]
        P1 = bar[-P1] => foo => bar
```
it will stall on the 2nd cycle point

**Comment by oliver-sanders** [comment]
> it will stall on the 2nd cycle point

Yep that's correct because `bar[-P1]` creates a dependence on the result of previous cycles. In the absence of this dependency there is no reason for the workflow to stall. Succeeded and failed are two possible outcomes, unless to create a dependence on one outcome or another there is nothing to stall.

There is the question of how to better enable users to design workflows which do "stall" or "fail" in the event of task failures where that is the desired solution.

This example works, however, it forces `bar` to run sequentially, which is not desirable:

```
P1 = bar[-P1] => bar
```

A dummy task hanging off of `bar` does the job better but convolutes the graph:

```
P1 = bar => dummy
# or better, some kind of forward trigger syntax to a dummy task in the final cycle
P1 = bar => dummy[$]
P1 = bar[*] => dummy
```

**Comment by MetRonnie** [comment]
There is currently a backwards compatibility in `runahead limit` where if you set it to a number `n`, it takes that to mean `PTnH`. I'm guessing we should remove that and not allow numbers anymore? We could take the number to mean `Pn` instead, but then that's a bit of a subtle change that could confuse users?

**Comment by hjoliver** [comment]
> There is the question of how to better enable users to design workflows which do "stall" or "fail" in the event of task failures where that is the desired solution.

That's an interesting one.  In Cylc 7, stalls were sometimes convenient, but happened for the wrong reasons (no choice due to pre-spawned waiting tasks with unsatisfied prerequisites). And it wasn't consistent (probably just consistent enough to make users think it was consistent).

In Cylc 8 the workflow doesn't need to stall unless there is literally nothing that can run under the circumstances.  Which is *the right think to do*. 

But maybe we do need a mechanism to allow deliberate planned stalls. One idea:

- if any task (or a particular task) fails, hold the workflow
- for individual tasks this could be done via an "action" in the graph (think of like a post-xtrigger that can affect scheduler behaviour): `foo:fail => @hold-scheduler`

**Comment by MetRonnie** [comment]
(Mainly a note to myself) I've just noticed: if you replace `P1 = foo => bar` with `R1 = foo => bar`, the workflow _does_ stall after `foo.1` fails. This seems to be because there aren't any more cycle points after that one; it won't stall on failure as long as there are still cycle points left to run. If you set `final cycle point = 2` for `P1 = foo => bar`, it will also stall after `foo.2` fails.

**Comment by hjoliver** [comment]
I wouldn't call that a "stall" in the traditional (Cylc 7) sense.  It's more that there's nothing else to run (flow finished) *but* the scheduler won't shut down because there's unhandled failed tasks in the task pool (which we've provisionally decided not to remove automatically).

**Comment by hjoliver** [comment]
Closing this one as invalid. If we need a "deliberate stall" feature, as described above, we can raise a new issue for that.
